File: Postfix.java
import java.util.ArrayDeque;
import java.util.Deque;

public class Postfix {

    /**
     * 
     * @param c
     * @return precedenceedence of the character
     */
    // metodo para poder verificar la precedenceedencia del caracter a analizar
    public static int precedence(char c) {
        switch (c) {
            case '|':
                return 1;
            case '.':
                return 2;
            case '*':
            case '+':
                return 3;
        }
        return -1;
    }

    /**
     * 
     * @param infix
     * @return postfix
     */
    // metodo para poder convertir la infixresion infix a postfix
    public static String infixToPostfix(String infix) {

        String postfix = "";

        // initializing empty stack
        Deque<Character> stack = new ArrayDeque<Character>();

        // System.out.println(infix);

        for (int i = 0; i < infix.length(); ++i) {

            char c = infix.charAt(i);

            // si el caracter es una letra o digito se agrega al ressultado
            if (Character.isLetterOrDigit(c))
                postfix += c;

            // Si el caracter es '(', push al stack.
            else if (c == '(')
                stack.push(c);

            // Si el caracter es ')', pop y output del stack hasta que se encuentre un '('
            else if (c == ')') {
                while (!stack.isEmpty()
                        && stack.peek() != '(') {
                    postfix += stack.peek();
                    stack.pop();
                }

                stack.pop();
            } else // si ha encontrado un operador
            {
                while (!stack.isEmpty()
                        && precedence(c) <= precedence(stack.peek())) {

                    postfix += stack.peek();
                    stack.pop();
                }
                stack.push(c);
            }
        }

        // pop de todos los operadores del stack
        while (!stack.isEmpty()) {
            if (stack.peek() == '(')
                return "Invalid infixression";
            postfix += stack.peek();
            stack.pop();
        }

        return postfix;

    }

}

File: Proyecto1/Postfix.py
from collections import deque


class Postfix:
    '''
    Postfix: Es la clase que se utiliza para poder convertir la expresion a postfix, desde un infix. 
    Toma como parametro el regex en infix.
    Se guarda la precedencia de cada operador para poder construir el postifx de manera correcta.
    '''

    def __init__(self, infix):
        self.infix = infix

    '''
    definimos la precedencia de los operadores
    '''

    def precedence(self, c):
        if c == '|':
            return 1
        elif c == '.':
            return 2
        elif c == '*':
            return 3
        else:
            return -1

    '''
    pasamos el regex de infix a postifx
    '''

    def infixToPostfix(self):
        postfix = ""

        # creamos una pila
        stack = deque()

        # iteramos por cada uno de los caracteres del regex
        for i in range(len(self.infix)):
            c = self.infix[i]

            # si es un operando lo agregamos al postfix
            if c.isalnum():
                postfix += c
            # si es un parentesis izquierdo lo agregamos a la pila
            elif c == '(':
                stack.append(c)
            # si es un parentesis derecho sacamos todos los elementos de la pila hasta encontrar un parentesis izquierdo
            elif c == ')':
                while stack and stack[-1] != '(':
                    postfix += stack.pop()
                if stack and stack[-1] == '(':
                    stack.pop()
            # si es un operador
            else:
                while stack and self.precedence(c) <= self.precedence(stack[-1]):
                    postfix += stack.pop()
                stack.append(c)

        while stack:
            if stack[-1] == '(':
                return "Invalid Infix Expression"
            postfix += stack.pop()

        return postfix
